client-app/src/Components/DonatedItemsList.tsx

```diff
@@ -0,0 +1,527 @@
+    };
+
+    const handleProgramChange = (
+        event: React.ChangeEvent<HTMLSelectElement>,
+    ): void => {
+        setSelectedProgram(event.target.value);
+    };
+    const handleBarcodeClick = (itemId: number): void => {
+        const selectedItem = donatedItems.find(item => item.id === itemId);
+        if (selectedItem) {
+            setSelectedItemDetails({
+                ...selectedItem,
+                statuses: selectedItem.statuses || [],
+            });
+        }
+        setModalIsOpen(true);
+    };
+
+    const updatePrograms = async (): Promise<void> => {
+        try {
+            const response = await fetch(
+                `${process.env.REACT_APP_BACKEND_API_BASE_URL}/program`,
+                {
+                    method: 'PUT',
+                    headers: {
+                        'Content-Type': 'application/json',
+                    },
+                    body: JSON.stringify({
+                        itemIds: selectedItems,
+                        programId: parseInt(selectedProgram),
+                    }),
+                },
+            );
+
+            if (!response.ok) {
+                throw new Error('Failed to update programs');
+            }
+
+            await fetchDonatedItems();
+            setSelectedItems([]);
+            setAssignProgramClicked(false);
+        } catch (err) {
+            const error = err as Error;
+            console.error('Error updating programs:', error);
+            setError('Failed to update programs. Please try again.');
+        }
+    };
+
+    const handleFilterByItemName = (
+        event: React.ChangeEvent<HTMLSelectElement>,
+    ): void => {
+        if (!event.target.value) {
+            setFilteredItems([]);
+            return;
+        }
+        const filtered = donatedItems.filter(
+            item =>
+                item.itemType.toLowerCase() ===
+                event.target.value.toLowerCase(),
+        );
+        setFilteredItems(filtered);
+    };
+
+    const handleFilterByProgram = (
+        event: React.ChangeEvent<HTMLSelectElement>,
+    ): void => {
+        if (!event.target.value) {
+            setFilteredItems([]);
+            return;
+        }
+        const programId = parseInt(event.target.value);
+        const filtered = donatedItems.filter(
+            item => item.programId === programId,
+        );
+        setFilteredItems(filtered);
+    };
+
+    const handleFilterByStatus = (
+        event: React.ChangeEvent<HTMLSelectElement>,
+    ): void => {
+        if (!event.target.value) {
+            setFilteredItems([]);
+            return;
+        }
+        const filtered = donatedItems.filter(
+            item =>
+                item.currentStatus.toLowerCase() ===
+                event.target.value.toLowerCase(),
+        );
+        setFilteredItems(filtered);
+    };
+
+    const toggleAssignProgram = (): void => {
+        setAssignProgramClicked(!assignProgramClicked);
+    };
```

Comment:
Add error handling for the barcode download process to manage failures in html2canvas or any other issues